File: 4- Elementry Search/Linear Search/ImprovedLS.java
import java.util.ArrayList;

public class ImprovedLS {
    public static <T extends Comparable<T>> int Search(ArrayList<T> iterable, T key) {
        int left = 0;
        int length = iterable.size();
        int right = length - 1;
        int position = -1;

        for (left = 0; left <= right; ) {
            if (iterable.get(left).equals(key)) {
                position = left;
                return position;
            }
            if (iterable.get(right).equals(key)) {
                position = right;
                return position;
            }
            left++;
            right--;
        }
        return -1;
    }
}

File: Sorts & Search & Shuffling Algorithms/Elementry Search/Linear Search/ImprovedLS.py
class ImprovedLS:
    @staticmethod
    def search(iterable, key):
        left = 0
        length = len(iterable)
        position = -1
        right = length - 1

        for left in range(0, right, 1):
            if iterable[left] == key:
                position = left
                return position
            if iterable[right] == key:
                position = right
                return position
            left += 1
            right -= 1

        return -1
